netebpfext/net_ebpf_ext.h

```diff
@@ -88,42 +112,63 @@ typedef struct _net_ebpf_ext_wfp_filter_id
     wchar_t* name;
     uint64_t id;
     net_ebpf_ext_wfp_filter_state_t state;
+    NTSTATUS error_code;
 } net_ebpf_ext_wfp_filter_id_t;
 
 typedef struct _net_ebpf_extension_wfp_filter_context
 {
-    volatile long reference_count;                                ///< Reference count.
-    const struct _net_ebpf_extension_hook_client* client_context; ///< Pointer to hook NPI client.
+    LIST_ENTRY link;                   ///< Entry in the list of filter contexts.
+    volatile long reference_count;     ///< Reference count.
+    EX_SPIN_LOCK lock;                 ///< Lock to protect the client context array.
+    uint32_t client_context_count_max; ///< Maximum number of hook NPI clients.
+    _Guarded_by_(
+        lock) struct _net_ebpf_extension_hook_client** client_contexts; ///< Array of pointers to hook NPI clients.
+    _Guarded_by_(lock) uint32_t client_context_count;                   ///< Current number of hook NPI clients.
+    const struct _net_ebpf_extension_hook_provider* provider_context;   ///< Pointer to provider binding context.
 
     net_ebpf_ext_wfp_filter_id_t* filter_ids; ///< Array of WFP filter Ids.
     uint32_t filter_ids_count;                ///< Number of WFP filter Ids.
 
-    bool client_detached : 1; ///< True if client has detached.
+    size_t client_data_size; ///< Size of client data associated with the filter context.
+    uint8_t* client_data;    ///< Client data associated with the filter context.
```

Comment:
Can you mention `attach parameter` in the description? Also, shouldn't `ebpf_extension_data_t` be used here?

Reply:
Removed this completely, it was not needed.